compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/AssertionLowering.kt

```diff
@@ -0,0 +1,177 @@
+                classSymbol.typeWith(), classSymbol,
+                CrIrType(asmType)
+            )
+        }
+    }
+
+    override fun lower(irFile: IrFile) {
+        // In legacy mode we treat assertions as inline function calls
+        if (context.state.assertionsMode != JVMAssertionsMode.LEGACY)
+            irFile.transformChildren(this, null)
+    }
+
+    override fun visitClass(declaration: IrClass, data: ClassInfo?): IrStatement {
+        val info = ClassInfo(declaration)
+
+        super.visitClass(declaration, info)
+
+        // Note that it's necessary to add this member at the end of the class, behind all user-visible initializers,
+        // since the Java language specification demands that assertions should run during class initialization.
+        // That is also the reason why this field records whether assertions are disabled rather than enabled.
+        // During initialization, $assertionsDisabled is going to be false, meaning that assertions are checked.
+        if (info.needsAssertionDisabledField) {
+            declaration.addMember(info.assertionsDisabledField)
+
+            // Some parents of local declarations are not updated during ad-hoc inlining
+            // TODO: Remove when generic inliner is used
+            declaration.patchDeclarationParents(declaration.parent)
+        }
+
+        return declaration
+    }
+
+    override fun visitCall(expression: IrCall, data: ClassInfo?): IrElement {
+        val function = expression.symbol.owner
+        if (!function.isAssert)
+            return super.visitCall(expression, data)
+
+        val mode = context.state.assertionsMode
+        if (mode == JVMAssertionsMode.ALWAYS_DISABLE)
+            return IrGetObjectValueImpl(
+                expression.startOffset,
+                expression.endOffset,
+                context.irBuiltIns.unitType,
+                context.irBuiltIns.unitClass
+            )
+
+        context.createIrBuilder(expression.symbol).run {
+            at(expression)
+            val assertCondition = expression.getValueArgument(0)!!
+            val lambdaArgument = if (function.valueParameters.size == 2) expression.getValueArgument(1) else null
+
+            return if (mode == JVMAssertionsMode.ALWAYS_ENABLE) {
+                checkAssertion(assertCondition, lambdaArgument)
+            } else {
+                require(mode == JVMAssertionsMode.JVM && data != null)
+                irIfThen(
+                    context.irBuiltIns.unitType,
+                    irNot(data.getAssertionDisabled(this)),
+                    checkAssertion(assertCondition, lambdaArgument)
+                )
+            }
+        }
+    }
+
+    private fun IrBuilderWithScope.checkAssertion(assertCondition: IrExpression, lambdaArgument: IrExpression?) =
+        irBlock {
+            val lambda = lambdaArgument?.asSimpleLambda()
+            val invokeVar = if (lambda == null && lambdaArgument != null) irTemporary(lambdaArgument) else null
```

Comment:
I don't think you need a temporary variable here.

Reply:
Because there can be side effects when evaluating the argument to a function, see `compiler/testData/codegen/box/assert/jvm/LambdaNotEvaluated.kt`.